src/drivers/imu/mpu6000/MPU6000.cpp

```diff
@@ -0,0 +1,989 @@
+		_px4_gyro.set_device_type(DRV_GYR_DEVTYPE_ICM20608);
+		break;
+
+	case MPU_DEVICE_TYPE_ICM20689:
+		_px4_accel.set_device_type(DRV_ACC_DEVTYPE_ICM20689);
+		_px4_gyro.set_device_type(DRV_GYR_DEVTYPE_ICM20689);
+		break;
+	}
+}
+
+MPU6000::~MPU6000()
+{
+	/* make sure we are truly inactive */
+	stop();
+
+	/* delete the perf counter */
+	perf_free(_sample_perf);
+	perf_free(_measure_interval);
+	perf_free(_bad_transfers);
+	perf_free(_bad_registers);
+	perf_free(_reset_retries);
+	perf_free(_duplicates);
+}
+
+int
+MPU6000::init()
+{
+
+#if defined(USE_I2C)
+	use_i2c(_interface->get_device_bus_type() == Device::DeviceBusType_I2C);
+#endif
+
+	/* probe again to get our settings that are based on the device type */
+	int ret = probe();
+
+	/* if probe failed, bail now */
+	if (ret != OK) {
+		PX4_DEBUG("CDev init failed");
+		return ret;
+	}
+
+	/* do init */
+	ret = CDev::init();
+
+	/* if init failed, bail now */
+	if (ret != OK) {
+		PX4_DEBUG("CDev init failed");
+		return ret;
+	}
+
+	if (reset() != OK) {
+		return ret;
+	}
+
+	return ret;
+}
+
+int MPU6000::reset()
+{
+	// if the mpu6000 is initialized after the l3gd20 and lsm303d
+	// then if we don't do an irqsave/irqrestore here the mpu6000
+	// frequently comes up in a bad state where all transfers
+	// come as zero
+	uint8_t tries = 5;
+	irqstate_t state;
+
+	while (--tries != 0) {
+		state = px4_enter_critical_section();
+
+		// Hold off sampling for 60 ms
+		_reset_wait = hrt_absolute_time() + 60000;
+
+		write_reg(MPUREG_PWR_MGMT_1, BIT_H_RESET);
+		up_udelay(10000);
+
+		// Wake up device and select GyroZ clock. Note that the
+		// MPU6000 starts up in sleep mode, and it can take some time
+		// for it to come out of sleep
+		write_checked_reg(MPUREG_PWR_MGMT_1, MPU_CLK_SEL_PLLGYROZ);
+		up_udelay(1000);
+
+		// Enable I2C bus or Disable I2C bus (recommended on data sheet)
+		write_checked_reg(MPUREG_USER_CTRL, is_i2c() ? 0 : BIT_I2C_IF_DIS);
+
+		px4_leave_critical_section(state);
+
+		if (read_reg(MPUREG_PWR_MGMT_1) == MPU_CLK_SEL_PLLGYROZ) {
+			break;
+		}
+
+		perf_count(_reset_retries);
+		px4_usleep(2000);
+	}
+
+	// Hold off sampling for 30 ms
+
+	state = px4_enter_critical_section();
+	_reset_wait = hrt_absolute_time() + 30000;
+	px4_leave_critical_section(state);
+
+	if (read_reg(MPUREG_PWR_MGMT_1) != MPU_CLK_SEL_PLLGYROZ) {
+		return -EIO;
+	}
+
+	px4_usleep(1000);
+
+	// SAMPLE RATE
+	_set_sample_rate(1000);
+	px4_usleep(1000);
+
+	_set_dlpf_filter(MPU6000_DEFAULT_ONCHIP_FILTER_FREQ);
+
+	if (is_icm_device()) {
+		_set_icm_acc_dlpf_filter(MPU6000_DEFAULT_ONCHIP_FILTER_FREQ);
+	}
+
+	px4_usleep(1000);
+	// Gyro scale 2000 deg/s ()
+	write_checked_reg(MPUREG_GYRO_CONFIG, BITS_FS_2000DPS);
+	px4_usleep(1000);
+
+	// correct gyro scale factors
+	// scale to rad/s in SI units
+	// 2000 deg/s = (2000/180)*PI = 34.906585 rad/s
+	// scaling factor:
+	// 1/(2^15)*(2000/180)*PI
+	_px4_gyro.set_scale(0.0174532f / 16.4f);//1.0f / (32768.0f * (2000.0f / 180.0f) * M_PI_F);
```

Comment:
Magic numbers? Do we want to clean up the code or just implement the new architecture? It does seem like a hassle, especially since this is pretty legacy at this point